includes/core/classes/endpoints/class-endpoint.php

```diff
@@ -0,0 +1,513 @@
+		$rewrite_base = $this->type_object->rewrite['slug'];
+		$slugs        = join( '|', $this->get_slugs() );
+		// Build the regular expression pattern for matching the custom endpoint URL structure.
+		$reg_ex_pattern = sprintf(
+			$this->reg_ex,
+			$rewrite_base,
+			$slugs
+		);
+		// Define the URL structure for handling matched requests via query vars.
+		// Example result: 'index.php?gatherpress_event=$matches[1]&gatherpress_ext_calendar=$matches[2]'.
+		$rewrite_url = add_query_arg(
+			$this->get_rewrite_atts(),
+			'index.php'
+		);
+
+		// Add the rewrite rule to WordPress.
+		add_rewrite_rule( $reg_ex_pattern, $rewrite_url, 'top' );
+
+		// Trigger a flush of rewrite rules.
+		$this->maybe_flush_rewrite_rules( $reg_ex_pattern, $rewrite_url );
+
+		// Allow the custom query variable by filtering the public query vars.
+		add_filter( 'query_vars', array( $this, 'allow_query_vars' ) );
+
+		// Handle whether to include a template or redirect the request.
+		add_action( 'template_redirect', array( $this, 'template_redirect' ) );
+	}
+
+	/**
+	 * Defines the rewrite replacement attributes for the custom feed endpoint.
+	 *
+	 * This method defines the rewrite replacement attributes
+	 * for the custom feed endpoint to be further processed by add_rewrite_rule().
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return array The rewrite replacement attributes for add_rewrite_rule().
+	 */
+	public function get_rewrite_atts(): array {
+		return array(
+			$this->type_object->name => '$matches[1]',
+			$this->query_var         => '$matches[2]',
+		);
+	}
+
+	/**
+	 * Creates a flag option to indicate that rewrite rules need to be flushed.
+	 *
+	 * This method checks if the generated rewrite rules already exist in the DB,
+	 * and if its rewrite URL matches the recently generated rewrite URL.
+	 * If any of this checks fail, the option to flush the rewrite rules will be set.
+	 *
+	 * This method DOES NO checks if the 'gatherpress_flush_rewrite_rules_flag' option
+	 * exists. It just adds the option and sets it to true. This flag
+	 * is being used to determine when rewrite rules should be flushed.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param  string $reg_ex_pattern The regular expression pattern for matching the custom endpoint URL structure.
+	 * @param  string $rewrite_url    The URL structure for handling matched requests via query vars.
+	 * @return void
+	 */
+	private function maybe_flush_rewrite_rules( string $reg_ex_pattern, string $rewrite_url ): void {
+		$rules = get_option( 'rewrite_rules' );
+
+		if ( ! isset( $rules[ $reg_ex_pattern ] ) || $rules[ $reg_ex_pattern ] !== $rewrite_url ) {
+			// Event_Setup->maybe_create_flush_rewrite_rules_flag // @todo maybe make this a public method ?!
+			// @see https://github.com/GatherPress/gatherpress/blob/3d91f2bcb30b5d02ebf459cd5a42d4f43bc05ea5/includes/core/classes/class-settings.php#L760C1-L761C63 .
+			add_option( 'gatherpress_flush_rewrite_rules_flag', true );
+		}
+	}
```

Comment:
This method could be simplified if `Event_Setup->maybe_create_flush_rewrite_rules_flag()` could maybe become a public method ?! Which in itself could be simplified, like I described in #880